### ISSUE TITLE
"fatal error: bad g status" and "fatal error: runtime: mcall called on m->g0 stack"

### ISSUE THREAD

**Issue by tomasz1986** [syncthing/syncthing#7784]
This comes from the same device as reported in https://github.com/syncthing/syncthing/issues/7764. This is the very first time that I'm seeing this kind of an error though.

```
Panic at 2021-06-22T09:17:09+09:00
fatal error: bad g status

runtime stack:
runtime.throw(0xf449a5, 0xc)
	runtime/panic.go:1117 +0x79
runtime.goschedImpl(0xc0004f8f00)
	runtime/proc.go:3325 +0x277
runtime.gopreempt_m(0xc0004f8f00)
	runtime/proc.go:3361 +0x3b
runtime.mcall(0x0)
	runtime/asm_amd64.s:327 +0x5e

goroutine 55449 [syscall, locked to thread]:
runtime.cgocall(0x365e80, 0xc0007e0ae0, 0x1000)
	runtime/cgocall.go:154 +0x55 fp=0xc000169108 sp=0xc0001690d0 pc=0x2f4775
syscall.Syscall6(0x7ffbcc432680, 0x5, 0x70c, 0xc0028ec000, 0x8000, 0xc00016921c, 0x0, 0x0, 0x0, 0x0, ...)
	runtime/syscall_windows.go:343 +0xf2 fp=0xc000169148 sp=0xc000169108 pc=0x361072
syscall.ReadFile(0x32ce6b, 0x11316a8, 0xc0004f8f00, 0xc000169378, 0x365c5b, 0x40000024, 0x365b80, 0x0)
	syscall/zsyscall_windows.go:1006 +0x105 fp=0xc0001691d8 sp=0xc000169148 pc=0x3bd265
runtime: unexpected return pc for syscall.Read called from 0x1
stack: frame={sp:0xc0001691d8, fp:0xc000169230} stack=[0xc000166000,0xc00016a000)
000000c0001690d8:  0000015000000000  0203000a00000000 
000000c0001690e8:  00000150452dffff  000000c0007e0800 
000000c0001690f8:  000000c000169138  0000000000361072 <syscall.Syscall6+242> 
000000c000169108:  0000000000365e80 <runtime.asmstdcall+0>  000000c0007e0ae0 
000000c000169118:  0000000000001000  01000150452c3400 
000000c000169128:  000000c0007e0800  00000000011318a0 
000000c000169138:  000000c0001691c8  00000000003bd265 <syscall.ReadFile+261> 
000000c000169148:  00007ffbcc432680  0000000000000005 
000000c000169158:  000000000000070c  000000c0028ec000 
000000c000169168:  0000000000008000  000000c00016921c 
000000c000169178:  0000000000000000  0000000000000000 
000000c000169188:  0000000000000000  0000000000000000 
000000c000169198:  0000000000000000  000000c00003d2f0 
000000c0001691a8:  000000c0028ec000  0000000000000000 
000000c0001691b8:  000000c00016921c  000000c0028ec000 
000000c0001691c8:  000000c000169220  00000000003b574f <syscall.Read+111> 
000000c0001691d8: <000000000032ce6b <runtime.asyncPreempt2+107>  00000000011316a8 
000000c0001691e8:  000000c0004f8f00  000000c000169378 
000000c0001691f8:  0000000000365c5b <runtime.asyncPreempt+219>  0000000040000024 
000000c000169208:  0000000000365b80 <runtime.asyncPreempt+0>  0000000000000000 
000000c000169218:  0000000000000000  0000000000000000 
000000c000169228: !0000000000000001 >000000c000169380 
000000c000169238:  000000c0001690f8  0000000000000000 
000000c000169248:  0000000000010244  000000c0001690d0 
000000c000169258:  000000c0007e0800  0000000b813ff858 
000000c000169268:  000000c0007e0858  0000000000000000 
000000c000169278:  0000000000000000  0000000000000000 
000000c000169288:  0000000000000000  0000000000000000 
000000c000169298:  0000000000000000  0000000000000000 
000000c0001692a8:  0000000000000000  0000000000000000 
000000c0001692b8:  0000000000000000  0000000000000000 
000000c0001692c8:  0000000000000000  6c75616665445c61 
000000c0001692d8:  74636e79737e5c74  6f6f437e676e6968 
000000c0001692e8:  706d742e7365696b  000000c00c6c8518 
000000c0001692f8:  0000000000000003  000000c00c6c85d0 
000000c000169308:  000000000000000a  000000c002684e40 
000000c000169318:  000000000000001a  000000c00c6c85e0 
000000c000169328:  000000000000000d 
syscall.Read(0xc000169380, 0xc0001690f8, 0x0, 0x10244, 0xc0001690d0, 0xc0007e0800, 0xb813ff858)
	syscall/syscall_windows.go:369 +0x6f fp=0xc000169230 sp=0xc0001691d8 pc=0x3b574f
created by github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).pullerIteration
	github.com/syncthing/syncthing/lib/model/folder_sendrecv.go:265 +0x4c5
```

[panic-20210622-091709.log](https://github.com/syncthing/syncthing/files/6699775/panic-20210622-091709.log)

```
Panic at 2021-06-23T15:41:08+09:00
fatal error: runtime: mcall called on m->g0 stack

runtime stack:
github.com/syncthing/syncthing/lib/ignore.(*Matcher).Hash(0xc001aaab40, 0x2f6287, 0xc0001b8780)
	github.com/syncthing/syncthing/lib/ignore/ignore.go:326
github.com/syncthing/syncthing/lib/model.(*folder).scanSubdirs.func1(0xc00025b500, 0xc001bfc000, 0x40)
	github.com/syncthing/syncthing/lib/model/folder.go:432 +0x59
github.com/syncthing/syncthing/lib/model.(*folder).scanSubdirs(0xc00025b500, 0xc008809430, 0x1, 0x1, 0x0, 0x0)
	github.com/syncthing/syncthing/lib/model/folder.go:527 +0x7a8
github.com/syncthing/syncthing/lib/model.(*folder).Serve(0xc00025b500, 0x1241578, 0xc000ec4e40, 0x0, 0x0)
	github.com/syncthing/syncthing/lib/model/folder.go:209 +0xb89
github.com/thejerf/suture/v4.(*Supervisor).runService.func2(0xc0000c4360, 0xc000000027, 0x298b3803ac8, 0xc00025b500, 0x1241578, 0xc000ec4e40, 0xc000f3c020, 0xc000657ce0)
	github.com/thejerf/suture/v4@v4.0.1/supervisor.go:551 +0x7e
```

[panic-20210623-154107.log](https://github.com/syncthing/syncthing/files/6699824/panic-20210623-154107.log)

**Comment by AudriusButkevicius** [comment]
Corrupt executable/bad memory?

I don't think we can fix the go runtime blowing up with bad internal state, something is going badly wrong with the hardware.

**Comment by tomasz1986** [comment]
The hardware seems fine though. The system has been running completely stable so far. There have been no other errors reported in the OS, and I haven't observed anything peculiar about its behaviour either. The memory could be possibly faulty, although I doubt it, as it is soldered into the motherboard, and as such it would have likely experienced issues in the past already. The same Syncthing binary has been used on multiple other machines with no issues too, so I doubt that something is wrong with it.

The only unusual Windows-related setting that I have just realised was that due to being an all-in-one device, the OS was set to use [Modern Standby](https://docs.microsoft.com/windows-hardware/design/device-experiences/modern-standby), which can interfere with other software workings (citing Micosoft: **"Both software and hardware must be made ready for low-power operation"**). I have just disabled it, so maybe this will make a difference.
